internal/security/bootstrapper/command/setupacl/tokencleaner/tokencleaner.go

```
@@ -0,0 +1,253 @@
+// the deletion requires the ACL write permission and hence using bootstrap ACL token
+// the caller should calls this to clean up the old tokens before creating new tokens, and the caller could ingore
+// the returned error from the Scrubber as it is not that critical if unable to delete the old tokens
+func (tr *TokenCleaner) ScrubOldTokens(bootstrapACLToken string) error {
+	// if the base directory to clean does not even exist, nothing to be cleaned
+	if !tr.isBaseDirExist() {
+		tr.lc.Infof("the token base directory %s does not exist, skip scrubbing old tokens", tr.tokenBaseDirAbs)
+		return nil
+	}
+
+	if len(bootstrapACLToken) == 0 {
+		return errors.New("required bootstrap token is empty")
+	}
+
+	// walk through the base directory and found the matched token file names and
+	// read the token out of it and delete the token from Consul
+	walkErr := filepath.Walk(tr.tokenBaseDirAbs, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("found error while walking through token base directory %s: %v", tr.tokenBaseDirAbs, err)
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		fileName := filepath.Base(path)
+		if fileName != tr.tokenFileName {
+			return nil
+		}
+
+		token, err := tr.readTokenFromFile(path)
+		if err != nil {
+			tr.lc.Warnf("found read token from file error: %v", err)
+			return err
+		}
+
+		// need to lookup self to obtain the token's accessor as Consul's delete token API
+		// only take accessor ID
+		tokenSelf, err := tr.retrieveSelfToken(path, token)
+		if err != nil {
+			tr.lc.Warnf("found read self token error: %v", err)
+			return err
+		} else if tokenSelf == nil {
+			return nil
+		}
+
+		// purge Consul token from agent server
+		if err := tr.deleteToken(path, tokenSelf.AccessorID, bootstrapACLToken); err != nil {
+			tr.lc.Warnf("found delete token error: %v", err)
+			return err
```

Based on other comment (deleting a non-existent token may cause 500 error), we need to not "return" here.

---

you meant return nil here?

---

under what condition you want not to propagate the error?

---

Never propagate.  Best effort to delete.  OK if doesn't work.

---

ok, this is an standard interface from Golang's filepath.Walk so I think you meant to return nil here

---

in the command i've already ignore the error from tokenCleaner so that should already take care that right?

---

Right. Return nil.  Walk will ignore the error.  Got it.

---

fixed.